### DISCUSSION TITLE
Enforcing periodic boundary conditions

### DISCUSSION THREAD

**Discussion by tairoon1** [idaholab/moose#16419] [Q&A Modules: Solid mechanics] [answered]
Hi everyone,

I would like to enforce that the displacement of two nodes on opposite sides of my mesh to have **not** the same value, but actually a value that I would like to pass along, i.e. u^2 - u^1 = A, with u^2 and u^1 the values of opposite nodes and A is the value I want to give. Is it possible to use the Constraints System (https://mooseframework.inl.gov/syntax/Constraints/index.html) for that? Or how would I enforce such a constraint?
I have also seen the Global Strain System, but I am not too sure how to use it...

Thanks for your help.

**Comment by crswong888** [comment]
See the section on "Function Transform Periodic Boundaries" on the [AddPeriodicBCAction](https://mooseframework.inl.gov/source/actions/AddPeriodicBCAction.html) page. Will something like this work for you? I think you could just make u^1 your primary boundary and define the transform to the u^2 secondary boundary like: u^2 = A + u^1.

**Reply by tairoon1** [reply]
Thank you for the suggestion, but I don't seem to be able to make it run with that.

**Reply by crswong888** [reply]
You’re right. It’s not designed to solve your type of problem. Please see my comment below.

**Comment by jiangwen84** [comment]
@SudiptaBiswas

**Comment by crswong888** [comment, marked as answer]
@tairoon1 I whipped up a little something for you on my MOOSE app. I basically copy and pasted the code for [`LinearNodalConstraint`](https://mooseframework.inl.gov/source/constraints/LinearNodalConstraint.html) and modified it to solve the residual: R = u1 - u2 + A. You can find the source code here:

[`EqualValuePlusConstant.h`](https://github.com/crswong888/scorpion/blob/master/include/constraints/EqualValuePlusConstant.h)
[`EqualValuePlusConstant.C`](https://github.com/crswong888/scorpion/blob/master/src/constraints/EqualValuePlusConstant.C)

To use `EqualValuePlusConstant` (lack of a better name), just copy and paste these files into your own MOOSE app and be sure to change `registerMooseObject("YourAppName", EqualValuePlusConstant);` to reflect "YourAppName" in the `.C` file. 

The object works pretty well, but proceed at your own risk. I didn't put very much effort into writing the code - just wanted to demonstrate how you, yourself, could begin to solve your problem with MOOSE. I'm not sure what sort of physics you are working with, but I created two fairly simple tests: one that solves a diffusion problem and another that solves a solid mechanics problem.

1. The diffusion problem: [`tairoon1_diffusion_constraint.i`](https://github.com/crswong888/scorpion/blob/master/inputs/users/tairoon1_diffusion_constraint.i), models a square domain with a scalar field, u, such that `u = 0` on the left and `u = 1` on the right by enforcing u_left = 0 and the constraint R = u_left - u_right + 1.

2. The mechanics problem: [`tairoon1_mechanical_constraint.i`](https://github.com/crswong888/scorpion/blob/master/inputs/users/tairoon1_mechanical_constraint.i), models a 12" x 12" x 12" steel cube that is fixed at the bottom. A displacement of 0.5" is applied in the x-direction to the nodes along the top left side of the cube, and one of 0.6" to the nodes along the top right side, by enforcing u_left = 0.5 and the constraint R = u_left - u_right + 0.1.

You can basically use my code as a template and create a more formal `MooseObject` for this simple multi-freedom constraint problem. Perhaps you even consider developing it in the Framework and opening a pull request :)

**Reply by tairoon1** [reply]
Thanks! This looks pretty good on first sight! I will take a look later today.

**Reply by tairoon1** [reply]
Just two questions: 1. Would it also be possible, to make the constant not a constant but a parsed function? Such that I can put A*x or A*y for example to make the displacement vary with coordinates?
2. I see that the constraint is enforced by a penalty approach. Would there also be any way of using Lagrange Multipliers in MOOSE? Has it been implemented in MOOSE?

**Reply by crswong888** [reply]
@tairoon changing A to be a function would be fairly simple. You just have to modify that `_constant` variable I created to be set by a function object that you can sample. Other than that the residual will be very much the same just slightly different syntax. You’ll want to create an input parameter with `params.addParam<FunctionName>()` and `getFunction()` and then sample it by pointing to its `value()`. There are plenty demonstrations of how this works on the repository.

I used a penalty formulation because that’s what I am most familiar with. I actually don’t really know how Lagrange multipliers work exactly, but I know there are examples of them in MOOSE. If you want to use Lagrange multipliers, you’ll have to figure out what the residual is and then just dig around the source code and look for a good example of Lagrange constraints.

**Reply by tairoon1** [reply]
Ok, thanks very much for all your help!

**Reply by tairoon1** [reply]
@crswong888 I followed your steps and tried to work it out for my problem. Some problems that I noticed so far are: it only ever seems to converge using PJFNK and not at all using Newton. Also there seems to be some problems whenever there are many constraints.

Could you please take a look at my [code](https://github.com/tairoon1/moosefiles)? There are the constraint files .h .C, and a mesh + input file. The example I am trying to compute is a simple shear of a unit cell with a stiff fiber inside. A displacement in x direction is applied on the top surface using constraints [topedge_x, topedge_y], while the right and left edge are constrained to be equal. However, the right and left edges behave overly stiff and the simulation does not look right. When I remove the constraints [rightedge_x, rightedge_y], the simulation result looks fine. 

Thank you!

**Comment by SudiptaBiswas** [comment]
If you are after periodic strain, you can check this out, https://mooseframework.inl.gov/moose/syntax/Modules/TensorMechanics/GlobalStrain/ 
There are some examples here, https://github.com/idaholab/moose/tree/next/modules/combined/examples/periodic_strain on how to set this up.
